### ISSUE TITLE
Bug: pop-up notifications in Chromium appear over i3lock screen

### ISSUE THREAD

**Issue by aliceriot** [i3/i3lock#22]
When screen is locked using i3lock new notifications from Chromium appear on the screen above i3lock and are readable by any passers-by.

This happens whenever Chromium is running, the screen is locked, and Chromium displays a pop-up notification.

Possible workarounds are just disabling Chromium notifications, but in my mind i3lock should be blocking stuff like this. I haven't done extensive testing to see if spawning other types of windows or window-like things (while the screen is locked) will result in similar behavior.

**Comment by Airblader** [comment]
Are you running a compositor?

**Comment by aliceriot** [comment]
Yes, I use compton (b/c I like shiny things). Could this be the source of the problem?

**Comment by Airblader** [comment]
Can you reproduce it without Compton?

**Comment by aliceriot** [comment]
No, if I kill compton the behavior doesn't happen. Any ideas on how to work around this? I'd like to keep drop shadows and notifications without leaking information to anyone who sees my locked screen.

**Comment by Airblader** [comment]
Sorry, but as stated in #8, we don't support any kind of compositing (though if you can come up with a PR, we certainly don't mind a fix for it), hence I'll close this issue.

I suggest you toy around with excluding any kind of compton effects on those popups and see if that changes something. Otherwise, try another lockscreen program. Also, if you do investigate this and can at least come up with the technical reason for why this happens, please don't hesitate sharing it with us – maybe a fix wouldn't be so hard.

**Comment by Airblader** [comment]
@aliceriot By the way, what are these popups? Is it actually Chromium? Or is it some notification daemon? If you happen to be running dunst, it allows pausing / resuming the daemon which you can wrap around the i3lock call to fix this issue. Other daemons might have similar functionality.

**Comment by aliceriot** [comment]
The popups are the chromium notification engine, gmail and such can send notifications which look much like desktop notifications using the framework. Technical document about them [here](https://developer.chrome.com/extensions/notifications). 

I'll definitely poke around at this! No promises on a PR though :D

Thanks for your help!

**Comment by Airblader** [comment]
If you just find out what the underlying problem is, without a PR, please share that too. Maybe we can figure something out then. :9

**Comment by Manouchehri** [comment]
I was having a similar issue with i3bar showing on top of i3lock (via a modifier key trigger) a few months ago. It's been fixed, not sure which commit though. 

Those two *might* be related, hopefully it'll give a starting place to look.

**Comment by Airblader** [comment]
> It's been fixed, not sure which commit though.

Run a `git bisect` to find out.

**Comment by Manouchehri** [comment]
Well yes, that would do it. I haven't ran across @aliceriot's issue with my setup yet, so I wasn't planning on fixing it; I was trying to give a bit of help. =)

**Comment by Airblader** [comment]
Finding out which commit "fixed" it would be helping, no need to come up with an actual fix (if it even exists) ;)

**Comment by Manouchehri** [comment]
Oh, I lied, I was able to trigger it again. compton still draws i3bar on top of i3lock. My bad, I thought it was fixed but I just disabled compton and forgot.

**Comment by sandsmark** [comment]
For reference, the solution seems to be to use xcb_composite_get_overlay_window() on the root window and using the result instead of just the root window for xcb_create_window() in open_fullscreen_window().

https://github.com/google/xsecurelock/commit/ee1825a4a113bd9a24bf8a499fe579d9a03d7612

**Comment by Stebalien** [comment]
A workaround is to set `unredir-if-possible` in compton.

**Comment by Cxarli** [comment]
For future readers: the solution provided above (`unredir-if-possible`) fixed this for me on Chromium/Chrome and Telegram Desktop notifications. However, by now `compton` has been renamed to `picom` so you'll have to edit `~/.config/picom/picon.conf`

**Comment by ouuan** [comment]
There should be something about `compton`/`picom` in the issue title, e.g. "Bug: pop-up notifications over i3lock screen when using compton".

I ran into this issue because I was using the fork i3lock-color, and I was not using any compositor. I found https://github.com/i3/i3lock/issues/22#issuecomment-290543752 and thought it meant I need to use `picom` to solve this issue. After setting up `picom`, this issue still persists. Then I found that this is actually a bug that is only in i3lock-color (Raymo111#210), not in i3lock.
